packages/core-bridge/src/helpers/abort_controller.rs

```diff
@@ -0,0 +1,153 @@
+    }
+}
+
+/// The inner state of an AbortController, shared between the Rust and JS sides.
+struct AbortControllerInner {
+    // The fact that we require a `Context` in `ensure_js_initialized` means that we are running on
+    // the Node's thread, which guarantees that there can't be multiple threads calling into that
+    // function concurrently; that should in theory aleviate the need to use a lock on `js_counterpart`.
+    //
+    // It is however possible for the rust-side controller to get aborted from a non-Node thread
+    // while the JS-side controller is being created on the Node thread, in which case we don't
+    // want the Rust-side thread to get blocked for the JS-side to complete instantiation.
+    //
+    // By modelizing the "JS initialization" and "is aborted" states as two distinct independant
+    // structures, we ensure that we're never blocking execution of either thread. This however
+    // means that either step may happen before the other, so we need to be careful not to miss
+    // sending the abort signal. The good news is that nothing bad will happen if we call the JS
+    // abort callback multiple times.
+    js_counterpart: OnceLock<Arc<AbortControllerJsCounterpart>>,
+    aborted: OnceLock<String>,
+}
+
+struct AbortControllerJsCounterpart {
+    signal: Root<JsObject>,
+    abort: JsCallback<(String,), ()>,
+}
+
+impl AbortControllerInner {
+    /// Create the JS AbortController if it hasn't been created yet.
+    /// Returns a reference to the signal object that can be passed to JS.
+    fn ensure_js_initialized<'cx, C: Context<'cx>>(
+        &self,
+        cx: &mut C,
+    ) -> BridgeResult<Handle<'cx, JsValue>> {
+        if let Some(js_counterpart) = self.js_counterpart.get() {
+            // Already initialized, return the signal
+            return Ok(js_counterpart.signal.to_inner(cx).upcast());
+        }
+
+        // Not initialized yet, create the JS AbortController
+        let global = cx.global_object();
+        let abort_controller_class = global.get::<JsFunction, _, _>(cx, "AbortController")?;
+
+        let abort_controller = abort_controller_class.construct(cx, [])?;
+        let signal = abort_controller.get::<JsObject, _, _>(cx, "signal")?;
+        let abort_fn = abort_controller.get::<JsFunction, _, _>(cx, "abort")?;
+
+        let abort_cb = JsCallback::new(cx, abort_fn, Some(abort_controller));
+
+        let js_counterpart = Arc::new(AbortControllerJsCounterpart {
+            signal: signal.root(cx),
+            abort: abort_cb,
+        });
+
+        let js_counterpart = match self.js_counterpart.set(js_counterpart.clone()) {
+            Ok(()) => {
+                // If the Rust controller has already been aborted, call the JS abort callback now
+                // VALIDATE: Do we need a memory barrier here to ensure that js_counterpart and aborted are coherent?
+                //           I assume that the get() call ensures visibility of the js_counterpart
```

Comment:
I believe this works, but it's really hard to be sure.

Another option, that I thiiiink works, is to use a mutex that protects both things. The only situation you need to deal with is having the JS call not happen under the lock - but that's easy, because it's already in an `Arc`. So, just clone it under the lock and then call it outside of it. We already accept that double-calls are fine.

That might make it a lot easier to reason about.